romidata/webcache.py

```diff
@@ -0,0 +1,172 @@
+from skimage.transform import resize
+import imageio
+import numpy as np
+
+def file_path(db, scanid, filesetid, fileid):
+    scan = db.get_scan(scanid)
+    fs = scan.get_fileset(filesetid)
+    f = fs.get_file(fileid)
+    return os.path.join(db.basedir, scan.id, fs.id, f.filename)
+
+def _hash(type, scanid, filesetid, fileid, size):
+    m = hashlib.sha1()
+    key = "%s|%s|%s|%s|%s" % (type, scanid, filesetid, fileid, size)
+    m.update(key.encode('utf-8'))
+    return m.hexdigest()
+
+# Image
+
+def image_hash(scanid, filesetid, fileid, size):
+    return _hash("image", scanid, filesetid, fileid, size)
+
+def image_resize(img, max_size):
+    i = np.argmax(img.shape[0:2])
+    if img.shape[i] <= max_size:
+        return img
+    if i == 0:
+        new_shape = [max_size, int(max_size * img.shape[1]/img.shape[0])]
+    else:
+        new_shape = [int(max_size * img.shape[0]/img.shape[1]), max_size]
+    return resize(img, new_shape)
+
+
+def image_cache(db, scanid, filesetid, fileid, size):
+    src = file_path(db, scanid, filesetid, fileid)
+    dir = os.path.join(db.basedir, scanid, "webcache")
+    os.makedirs(dir, exist_ok=True)
+    dst = os.path.join(dir, image_hash(scanid, filesetid, fileid, size))
+    
+    resolutions = {
+        "large": 1500,
+        "thumb": 150
+    }
+    maxsize = resolutions.get(size) 
+
+    image = imageio.imread(src)
+    # remove alpha channel
+    if image.shape[2] == 4:
+        image = image[:,:,:3]
+    cached_image = image_resize(image, maxsize)
+    imageio.imwrite(dst, cached_image, format="jpg")
+
+    print("Converted %s to %s, size %d" % (src, dst, maxsize))
+
+    return dst;
+
+
+def image_cached_path(db, scanid, filesetid, fileid, size):
+    path = os.path.join(db.basedir, scanid, "webcache",
+                        image_hash(scanid, filesetid, fileid, size))
+    if not os.path.isfile(path):
+        image_cache(db, scanid, filesetid, fileid, size)
+    return path
+
+
+def image_path(db, scanid, filesetid, fileid, size):
+    if size == "orig":
+        print("Using original file")
+        return file_path(db, scanid, filesetid, fileid)
+    else:
+        print("Using cached file")
+        return image_cached_path(db, scanid, filesetid, fileid, size)
+
+# PointCloud
+
+def pointcloud_hash(scanid, filesetid, fileid, size):
+    return _hash("pointcloud", scanid, filesetid, fileid, size)
+
+
+def pointcloud_resize(pointcloud, max_size):
+    try:
+        from open3d import open3d
+    except:
+        import open3d
+        
+    if len(pointcloud.points) < max_pointcloud_size:
+        return pointcloud
+
+    downsample = len(pointcloud.points) // max_size + 1
+    try:
+        return open3d.geometry.uniform_down_sample(pointcloud, downsample)
+    except:
+        return pointcloud.voxel_down_sample(downsample)
+
+
+def pointcloud_cache(db, scanid, filesetid, fileid, size):
+    try:
+        from open3d import open3d
+    except:
+        import open3d
```

Comment:
Too broad exception clause, should be replaced by:
```python
    try:
        from open3d import open3d
    except ModuleNotFoundError:
        msg = "Open3D is not installed, do so with: pip install open3d"
        raise ModuleNotFoundError(msg)
    except ImportError:
        import open3d
```
This also make a difference if `open3d` is installed or not!